### PR TITLE
feat: Extension pack for course

### LINKED ISSUE #648: Add Support to Buy Extension/Revision Pack

### PR DESCRIPTION
**Pull request by jddeep**
Fixes #648 

Changes: Extension pack for the course

Screenshots for the change:

### CODE REVIEW

File: app/src/main/java/com/codingblocks/cbonlineapp/mycourse/overview/OverviewFragment.kt [hunk 1]
```diff
@@ -49,9 +54,11 @@ class OverviewFragment : BaseCBFragment(), AnkoLogger {
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+        courseExtensionsRv.setRv(requireContext(), extensionListAdapter)
         viewModel.run?.distinctUntilChanged()?.observer(viewLifecycleOwner) { courseAndRun ->
             viewModel.runStartEnd = Pair(courseAndRun.runAttempt.end.toLong() * 1000, courseAndRun.run.crStart.toLong())
             viewModel.runId = (courseAndRun.run.crUid)
+            viewModel.fetchExtensions(courseAndRun.course.cid)
```
Comment (jddeep) [inline app/src/main/java/com/codingblocks/cbonlineapp/mycourse/overview/OverviewFragment.kt]: @aggarwalpulkit596 Is this the `productId` for a course?

Comment (aggarwalpulkit596) [reply app/src/main/java/com/codingblocks/cbonlineapp/mycourse/overview/OverviewFragment.kt]: no it will be the runId

**Comment by jddeep** [comment]
@aggarwalpulkit596 updated. Thanks

File: app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt [hunk 2]
```diff
@@ -113,6 +117,39 @@ class MyCourseViewModel(
         }
     }
 
+    fun fetchExtensions(productId: String?) {
+        runIO {
+            when (val response = productId?.let { repo.getExtensions(Integer.parseInt(it)) }) {
+                is ResultWrapper.GenericError -> setError(response.error)
+                is ResultWrapper.Success -> with(response.value) {
+                    this.enqueue(retrofitCallback { _, response ->
```
Comment (aggarwalpulkit596) [inline app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt]: again what are you trying to achieve here I'm not getting it you should be focusing on understanding the code first rather just completing the work

**Review by aggarwalpulkit596 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt (hunk 2 above)
Comment (jddeep) [reply app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt]: Again the same answer I would give here, being a network request am enquing with the callback to use the response. Let me know if I should use `awaitResponse().body()` directly without enquing (at least based on my understanding, am open to suggestions to learn from it)

File: app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt [hunk 3]
```diff
@@ -116,6 +121,38 @@ class MyCourseViewModel(
...
+        }
+    }
+
+    fun buyExtension(extensionId: Int?) {
+        runIO {
+            when (val response = extensionId?.let { repo.buyExtension(it) }) {
+                is ResultWrapper.GenericError -> setError(response.error)
+                is ResultWrapper.Success -> with(response.value) {
+                    if (isSuccessful) {
+                        purchase.postValue(true)
```
Comment (jddeep) [inline app/src/main/java/com/codingblocks/cbonlineapp/mycourse/MyCourseViewModel.kt:147]: @aggarwalpulkit596 bhaiya, for buying functionality, we can observe `purchase` in the Overview fragment and do proceedings. But what should be the flow from there onwards? Please let me know bhaiya.

**Comment by aggarwalpulkit596** [comment]
@jddeep any update on this

**Comment by jddeep** [comment]
@aggarwalpulkit596 Actually have been caught up in something lately. Haven't got time to sit down and update the PRs. But am hoping to get this done by this weekend.

**Comment by jddeep** [comment]
@aggarwalpulkit596 Here are some of the screenshots after my work done today - 
exts

But after pulling the latest updates from upstream, there seem to be a hell lot of conflicts. And Android Studio doesn't have options like Accept Incoming change, etc. So it seems like I might need to make the PR from another branch as I don't git that much patience to manually, fix the conflicts :sweat_smile: 
Let me know your thoughts.

File: app/src/main/java/com/codingblocks/cbonlineapp/mycourse/overview/OverviewFragment.kt (hunk 1 above)
Comment (jddeep) [reply app/src/main/java/com/codingblocks/cbonlineapp/mycourse/overview/OverviewFragment.kt]: Also, this seems to give the desired results for `course.productId`

**Comment by aggarwalpulkit596** [comment]
Yeah it looks good to me, some minor changes are required for the UI rest looks good but I'm still worried about how do you select an item and enable the buy extension button in that case

**Comment by aggarwalpulkit596** [comment]
@jddeep are you still working on this one ?

**Comment by jddeep** [comment]
@aggarwalpulkit596 I am thinking to use the `selectionTracker` methodology for each card tap and then proceeding to the checkout activity.

**Comment by aggarwalpulkit596** [comment]
@jddeep can you wind up this one by end of the week need this on a priority basis

**Comment by jddeep** [comment]
@aggarwalpulkit596 surely will try :+1:

**Comment by aggarwalpulkit596** [comment]
@jddeep any updates on this?

**Comment by jddeep** [comment]
@aggarwalpulkit596 Yes, but after linking with the selectionTracker, the app was crashing whenever I tried to scroll across. Also, the logcat seems to get a reset once, the app crashes and fills up with new logs as the app restarts upon crash. So, it is difficult for me to figure out what could be the possible error :confused: Any ideas?

**Comment by aggarwalpulkit596** [comment]
Use a debugger or add some logs to find the crash reason and tbh finding out crash logs is a trivial thing there's no rocket science in that.

**Comment by jddeep** [comment]
@aggarwalpulkit596 Yes ik, it is meant to be simple af. I have added multiple logs as well. The problem is that, once the app crashes, all the logs from the logcat are cleared automatically and starts afresh cause the app restarts automatically after crash..

**Comment by aggarwalpulkit596** [comment]
@jddeep if you are still working on this lmk otherwise I'll make a PR myself by tommorow

**Comment by jddeep** [comment]
@aggarwalpulkit596 Please go ahead. Seems like I am in the middle of some stuff. Apologies on the delay. Really wanted to finish this off as had done a lot of work already :/

**Pull request closed without merging**